biz.ganttproject.core/src/main/java/biz/ganttproject/core/chart/scene/TaskTableSceneBuilder.kt

```diff
@@ -0,0 +1,91 @@
+package biz.ganttproject.core.chart.scene
+
+import biz.ganttproject.core.chart.canvas.Canvas
+import biz.ganttproject.core.chart.canvas.TextMetrics
+import java.awt.Color
+import java.lang.Integer.max
+
+class TaskTableSceneBuilder(
+  private val rows: List<RowContent>,
+  private val input: InputApi,
+  val canvas: Canvas = Canvas()
+) {
+  private val dimensions = calculateDimensions()
+
+  fun build() {
+    canvas.clear()
+    val rectangle = canvas.createRectangle(0, 0, dimensions.width, dimensions.height)
+    rectangle.backgroundColor = Color.WHITE
+    paintTasks(rows, PaintState(input.horizontalOffset, 0, 0))
+  }
+
+  private fun calculateDimensions(rows: List<RowContent>? = null): Dimension {
+    var height = 0
+    var width = 0
+
+    (rows ?: this.rows).forEach {
+      height += input.rowHeight
+      width = max(width, it.getWidth(input.textMetrics))
+      if (it is Task) {
+        val subtasksDimensions = calculateDimensions(it.subrows)
+        height += subtasksDimensions.height
+        width = max(width, subtasksDimensions.width + input.depthIndent)
+      }
+    }
+
+    if (rows == null) {
+      width += 2 * input.horizontalOffset
+    }
+    return Dimension(height, width)
+  }
+
+  private fun paintTasks(rows: List<RowContent>, initState: PaintState): PaintState {
+    var state = initState
+    rows.forEach {
+      val rectangle = canvas.createRectangle(
+        0, state.y, dimensions.width, input.rowHeight
+      )
+      if (state.rowNumber % 2 == 1) {
+        rectangle.style = "odd-row"
+      }
+
+      if (it is Task) {
+        paintTask(it, state.x, rectangle.middleY)
+        val subtasksState = paintTasks(it.subrows, state.toNextRow(input.depthIndent))
+        subtasksState.x = state.x
+        state = subtasksState
+      } else {
+        state = state.toNextRow(0)
+      }
+    }
+    return state
+  }
+
+  private fun paintTask(task: Task, x: Int, y: Int) {
+    val text = canvas.createText(x, y, task.name)
+    text.setAlignment(Canvas.HAlignment.LEFT, Canvas.VAlignment.CENTER)
+  }
+
+  interface RowContent {
```

Comment:
maybe make more simple input interface, like this (pseudo-code):
```
RowContent {
  width: Int
  indent: Int
  subrows: List<RowContent>
}
```

TextMetrics can be encapsulated into RowContent implementation (it must be the same for all rows and it is obviously the same as one in InputApi)
Indents and child rows may be needed not only for tasks (e.g. this code can also be used for drawing resources and assignments)

I am also not sure if we really need the concept of subrows. Maybe just let the calling side build the list which is ordered in depth first order with the appropriate indents?